src/esp_part_dump.erl

```diff
@@ -0,0 +1,266 @@
+        end,
+    Partition = lookup_partition_by_offset(Offset, Partitions),
+    Partition.
+
+% @private
+-spec lookup_partition_by_offset(Offset :: non_neg_integer(), Partitions :: binary()) ->
+    {Name :: string(), Type :: atom()}
+    | {error, {Name :: list(), {Reason :: any(), Type :: atom() | byte()}}}.
+lookup_partition_by_offset(Offset, Partitions) ->
+    PartitionsData = parse_data_partitions(Partitions),
+    case lists:keyfind(Offset, 1, PartitionsData) of
+        false ->
+            {error, {"none", "No partition aligned to address"}};
+        {Offset, Name, protected} ->
+            {error, {Name, {not_data_partition, protected}}};
+        {Offset, Name, {reserved, Type}} ->
+            {error, {Name, {reserved, Type}}};
+        {Offset, Name, Type} ->
+            {Name, Type}
+    end.
+
+% @private
+-spec get_partition_data(Esptool :: string(), Port :: string(), TempFile :: string()) ->
+    PartitionData :: binary().
+get_partition_data(Esptool, Port, TempFile) ->
+    case file:read_file(TempFile) of
+        {ok, Data} ->
+            Data;
+        {error, enoent} ->
+            dump_device_partition(Esptool, Port, TempFile)
+    end.
+
+%% @private
+-spec dump_device_partition(Esptool :: string(), Port :: string(), TempFile :: string()) ->
+    PartitionData :: binary().
+dump_device_partition(Esptool, Port, TempFile) ->
+    BaseArgs = [
+        "read_flash",
+        "0x8000",
+        "0xC00",
+        TempFile
+    ],
+    Args =
+        case Port of
+            "auto" ->
+                BaseArgs;
+            _ ->
+                ["--port", Port | BaseArgs]
+        end,
+    case os:getenv("TEST") of
+        false ->
+            ok;
+        _ ->
+            rebar_api:info("~s ~s", [Esptool, lists:flatten(lists:join(" ", Args))])
+    end,
+
+    Tool =
+        try
+            open_port({spawn_executable, Esptool}, [
+                {args, Args}, stderr_to_stdout, use_stdio, exit_status
+            ])
+        catch
+            error:enoent ->
+                error("esptool.py not found");
+            error:eacces ->
+                error(not_executable, Esptool)
+        end,
+    receive
+        {Tool, {exit_status, 0}} ->
+            ok;
+        {Tool, {exit_status, Status}} ->
+            error({"esptool.py failure", Status})
+    after 120000 ->
+        error("esptool.py failed (2 minute timeout exceeded)")
+    end,
+    Partition_data =
+        case file:read_file(TempFile) of
+            {ok, Data} ->
+                Data;
+            {error, enoent} ->
+                error(no_device_dump);
+            Error ->
+                error(Error)
+        end,
+    Partition_data.
+
+%% @private
+-spec parse_data_partitions(PartitionTable :: binary()) ->
+    [
+        {
+            Offset :: non_neg_integer(),
+            Name :: string(),
+            Type ::
+                atom() | byte() | protected | {Type :: atom() | byte() | protected, Reason :: any()}
```

Comment:
Instead of `atom()` and `any()`, we could list types currently returned by this function (`phy | avm_app | undefined | byte() | protected | {reserved, ota | nvs | coredump | nvs_keys | fat | efuse | spiffs | littlefs}`)